fe/fe-core/src/main/java/com/starrocks/sql/analyzer/Scope.java

```diff
@@ -40,6 +41,8 @@ public class Scope {
     private List<PlaceHolderExpr> lambdaInputs = Lists.newArrayList();
 
     private boolean isLambdaScope = false;
+    private List<String> userVariableDependencyWithoutFind = new ArrayList<>();
```

Comment:
can you explain why you add this variable to the scope?

Reply:
+1

Reply:
purpose:When using CheckUserVariableDependencyVisitor to check whether UserVariable.unevaluatedExpression depends on other UserVariable, use Scope to record the variables that unevaluatedExpression depends on. Then analyzeUserVariableExprDependency method can return the record to UserVariable Object。

**the UserVariable.evaluatedExpression value is inaccurate in the follwing sql**
1 set @a = 6, @b = @a + 1, @c = @b + 1;
1.1）after SetStmtAnalyzer analyze the sql
var a's evaluatedExpression is IntLiteral(5)
var b's evaluatedExpression is NullLiteral
var c's evaluatedExpression is NullLiteral
1.2)after SetExecutor
ConnectContext.userVariables will put var a, b, c. then result is 5, NULL, NULL(error result)

so In step 1.2, **SetExecutor determines whether it is necessary to re-analyze UserVariable by userVariableDependencyWithoutFind.** Through re-analyze op, the result is 5,6,7.

purpose:To minimize calls to re-analyze, record only dependencies that don't exist. so the var is _withoutFind_
set @a=@e + 1,the sql dont execute re-analyze.

Reply:
Not using scope to record non-existent user variables
@imay @HangyuanLiu